### PR TITLE
Document rules internals - v1

### PR DESCRIPTION
**Pull request by jufajardini**
While advancing on the task of better understanding and describing how Suricata inspects packets and trying to explain to Jason what I did understand, we thought that adding this documentation could be a good intermediary step...

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/5449

Describe changes:
Added a page that explains how rules are prioritized by Suri, as well
as what main different types of inspection happen and what elements are
involved when ordering rules.

### CODE REVIEW

File: doc/userguide/rules/rules-internals.rst [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+--------
+
+Rules are provided to Suricata via rules files. Starting from those, the
+Detection Engine loader will:
+
+#. load all the signatures;
+#. check for validity (non-existing keywords; duplicated sid etc);
+#. report stats on loaded, good and bad signatures;
+#. sort all valid signatures and store them in a list in the Detection Engine
+   Contect (``DetectEngineCtx``) structure, taking into consideration several
```
Comment (jmtaylor90) [inline doc/userguide/rules/rules-internals.rst:37]: Is this Detect Engine Context?

**Comment by jmtaylor90** [comment]
This is great, thank you!

File: doc/userguide/rules/rules-internals.rst [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+Inspection Flow
+---------------
+
+Once it is time to inspect network traffic against the loaded rules, the
+Detection Engine will match against - if applicable:
+
+#. IPOnly rules (those where packet payload isn't inspected);
+#. Packet/payload related rules;
+#. Frame keywords;
+#. Application layer protocol transactions.
```
Comment (victorjulien) [inline doc/userguide/rules/rules-internals.rst:136]: This deserves a bit more explanation, as it has its own logic. Each parser has its own state machine, where it uses a per direction parsing state "progress". For each progress value keywords can be registered. E.g. http has a value "request line available" for which we registered keywords like http.uri, http.method, etc. If in the traffic we reach this state, the sigs with those keywords may be already evaluated, even if they have a lower prio than say a body inspecting sig.

**Review by victorjulien — CHANGES_REQUESTED** [review]
Nice work. Some inline comments.

File: doc/userguide/rules/rules-internals.rst (hunk 1 above)
Comment (jufajardini) [reply doc/userguide/rules/rules-internals.rst:37]: Yes, that. Thanks!

**Comment by jufajardini** [comment]
Replaced by: https://github.com/OISF/suricata/pull/7655

**Pull request closed without merging**